### PR TITLE
First attempt at checkpoints

### PR DESCRIPTION
**Pull request by ThomasWMarshall**
## Description
Adds checkpoint saving.

## How Has This Been Tested?
Running locally, verifying that the saved files contain the expected information.

### CODE REVIEW

**Comment by ThomasWMarshall** [comment]
Both the ApolloResearch sample repo that this repo was forked from and the starter code found in `train_gpt2.py` contain functions for saving parameters. If we can, I suggest that we forgo the code in `train_gpt2.py` and standardize on the simpler method from the ApolloResearch repo which really boils down to `torch.save(model.state_dict(), model_file)`. This would simplify things by keeping them consistent between `train_llama.py` and `train_gpt2.py`.

File: simple_stories_train/train_gpt2.py [hunk 1]
```diff
@@ -57,6 +58,8 @@
 from torch.nn import functional as F
 from torch.nn.parallel import DistributedDataParallel as DDP
 
+from utils import save_model_and_config, print0, is_checkpoint_step
```
Comment (danbraunai) [inline simple_stories_train/train_gpt2.py:61]: Best to use package name to import, i.e. simple_stories_train.utils. Doing so lets you run code from anywhere without issues.

File: simple_stories_train/train_gpt2.py [hunk 2]
```diff
@@ -903,13 +898,19 @@ def get_lr(it):
...
     if args.output_dir:
         os.makedirs(args.output_dir, exist_ok=True)
         logfile = os.path.join(args.output_dir, "main.log")
         # create the log file "main.log" inside it, and wipe it clean
         with open(logfile, "w") as f:
             pass
 
+        # set our checkpoints directory and save off the initilized model
+        checkpoints_dir = Path(args.output_dir) / 'checkpoints'
+        os.makedirs(checkpoints_dir, exist_ok=True)
```
Comment (danbraunai) [inline simple_stories_train/train_gpt2.py:911]: You can do something like `checkpoints_dir.mkdir(parents=True, exist_ok=True)` and not have to mix Pathlib with os.

File: simple_stories_train/utils.py [hunk 3]
```diff
@@ -1,28 +1,30 @@
...
-from simple_stories_train.log import logger
+def print0(*args: Any, **kwargs: Any):
+    # modified print that only prints from the master process
+    # if this is not a distributed run, it's just a print
+    if int(os.environ.get("RANK", 0)) == 0:
+        print(*args, **kwargs)
 
 
-def save_model_and_config(config: BaseModel, save_dir: Path, model: nn.Module, epoch: int) -> None:
+def is_checkpoint_step(step: int):
```
Comment (danbraunai) [inline simple_stories_train/utils.py:15]: return type is a bool

File: simple_stories_train/train_gpt2.py [hunk 4]
```diff
@@ -1021,6 +1022,10 @@ def get_lr(it):
             with open(logfile, "a") as f:
                 f.write("s:%d trl:%f\n" % (step, lossf))
 
+        if checkpoints_dir is not None and is_checkpoint_step(step + 1):
+            # save checkpoint
```
Comment (danbraunai) [inline simple_stories_train/train_gpt2.py:1026]: Don't need comment here

File: simple_stories_train/train_gpt2.py [hunk 5]
```diff
@@ -1021,6 +1022,10 @@ def get_lr(it):
             with open(logfile, "a") as f:
                 f.write("s:%d trl:%f\n" % (step, lossf))
 
+        if checkpoints_dir is not None and is_checkpoint_step(step + 1):
```
Comment (danbraunai) [inline simple_stories_train/train_gpt2.py:1025]: I think is_checkpoint_step should just take in `step` and not `step + 1`

File: simple_stories_train/train_llama.py [hunk 6]
```diff
@@ -892,13 +888,19 @@ def get_lr(it):
...
     if args.output_dir:
         os.makedirs(args.output_dir, exist_ok=True)
         logfile = os.path.join(args.output_dir, "main.log")
         # create the log file "main.log" inside it, and wipe it clean
         with open(logfile, "w") as f:
             pass
 
+        # set our checkpoints directory and save off the initilized model
+        checkpoints_dir = Path(args.output_dir) / 'checkpoints'
+        os.makedirs(checkpoints_dir, exist_ok=True)
```
Comment (danbraunai) [inline simple_stories_train/train_llama.py:901]: os -> Pathlib.mkdir

File: simple_stories_train/utils.py [hunk 7]
```diff
@@ -1,28 +1,30 @@
...
 from pathlib import Path
 
 import torch
-import yaml
-from pydantic import BaseModel
 from torch import nn
+from typing import Any
 
-from simple_stories_train.log import logger
+def print0(*args: Any, **kwargs: Any):
```
Comment (danbraunai) [inline simple_stories_train/utils.py:8]: nit: return type annotation (None)

File: simple_stories_train/utils.py [hunk 8]
```diff
@@ -1,28 +1,30 @@
...
+    # modified print that only prints from the master process
+    # if this is not a distributed run, it's just a print
+    if int(os.environ.get("RANK", 0)) == 0:
+        print(*args, **kwargs)
 
 
-def save_model_and_config(config: BaseModel, save_dir: Path, model: nn.Module, epoch: int) -> None:
+def is_checkpoint_step(step: int):
+    return (0 < step < 1000 and (step & (step - 1)) == 0) or step % 1000 == 0
```
Comment (danbraunai) [inline simple_stories_train/utils.py:16]: bahaha I have never seen `(step & (step - 1)) == 0)` before. Ludicrous, but yeah it's nice. Though I would put a docstring here explaining what this actually does. Ideally I'd also add a simple test `test_checkpoint_step` which just runs this over some of the interesting values and gives you what you'd expect.

**Review by danbraunai — CHANGES_REQUESTED** [review]
Thanks for this! Made some comments.

**Comment by danbraunai** [comment]
>This would simplify things by keeping them consistent between train_llama.py and train_gpt2.py.

I'm not sure we care much about supporting gpt2 for now. But yeah as to your general point, the llmc code that this script was ~copied from has the property that nearly everything is in one file. This can be nice. But yeah for our purposes I don't think this is necessary so I'm fine with starting to split things up into new files (as you have done with your utils).

**Comment by ThomasWMarshall** [comment]
Thanks for the review! I've done my best to address them all. Please don't hesitate to reopen any of them if you'd like me to revisit.

**Review by danbraunai-apollo — APPROVED** [review]
Looks great, cheers. I'll merge it.

**Pull request merged**